### ISSUE TITLE
change the raku spec versioning

### ISSUE THREAD

**Issue by eiro** [perl6/problem-solving#141]
for someone who don't know the raku history, there is no way to understand that 6.e is actually the 3rd version of the spec.

so i really suggest raku3.

**Comment by AlexDaniel** [comment]
I've said it before and I'll say it again: the way we do versioning right now totally made sense when the language was called Perl 6. But with the proper name that we finally have now our current versioning is borderline stupid. We can't go back to 3, but I believe we can go to 7.0.0 or higher and drop the lettering. This was discussed fairly recently, but there wasn't a consensus or anything like that yet.

**Comment by jnthn** [comment]
> We can't go back to 3, but I believe we can go to 7.0.0 or higher and drop the lettering

I was about to write "we'd probably have to go to 7 to have the sorting work out correctly", and you beat me to it. I'm open to this, though would be interested to see further input.

**Comment by FCO** [comment]
if `e` is higher than `6`, I'd prefer version `e`

**Comment by AlexDaniel** [comment]
> if `e` is higher than `6`, I'd prefer version `e`

We can prefer it in our bubble but everyone else is still confused about the letter. We didn't release 6.0 so a lot of people didn't realize that 6.c or 6.d means that we already had a proper language release. Raku 7.0.0 (or 7.0?) can be a great marketing potential.

**Comment by FCO** [comment]
I don't like the 7 because there were no 7 versions...

**Comment by duncand** [comment]
I agree that the 6.x versioning scheme is quite awkward and we should get away from that.

My preference is to actually go back to version 1.0 instead of 6.e.

A precedent for this is when Objective-C was renamed to Swift, the latter went to version 1.0.

Since the old versions all have letters, and if the new versions don't have letters, we can special case for backwards compatibility that any 6.c or 6.d are ordered before 1.0 or any higher numbers we eventually get to.

If that idea of going to 1.0 is considered unpalatable, then I support 7.0 instead of 6.e, not because it is the 7th version, but because it is the next version after the 6.x we had before and so sorting would work right under unified rules.

Major version numbers are never supposed to indicate this is the Nth version, it is only about relative rank between existing versions.

A key thing we need to account for however is that we need a distinct version namespace from Perl 5, so that either that or Raku can use 1.0 or 7.0 or 30.0 or whatever and it would be trivial for a compiler to be given source in either language and be able to know what language it was given.

**Comment by FCO** [comment]
Sorry, I may be wrong... but Swift and Objective-C are 2 different languages...

**Comment by duncand** [comment]
Well Apple was certainly treating Swift 1 as the next version of what used to be called Objective C, a total substitution.  Does anyone know otherwise, that newer versions of Objective C continued to be made and used by Apple after Swift came out?

**Comment by vrurg** [comment]
@duncand That worked for Swift because there was no exiting code base to maintain compatibility with. We do have that base which already relies on `6.c`/`6.d`.

I'm not a big fan of changing the versioning scheme. But since the overall support for this move is growing, I'd rather upvote for `v7.0` option. 

For those who worried about "there was no previous versions" issue – try finding Windows 6? Or Java 2,3,4? Can't recall right now, but I certainly knew an application which started from something higher than 1.x.

More than skipping, I'm worried about the syntax we're using. Because `use v7;` might be Perl's statement too. We need something more explicit, like `use raku;`, `use raku:ver<7>;`, etc.

**Comment by duncand** [comment]
> More than skipping, I'm worried about the syntax we're using. Because `use v7;` might be Perl's statement too. We need something more explicit, like `use raku;`, `use raku:ver<7>;`, etc.

I completely agree, this is the biggest issue.

**Comment by lizmat** [comment]
I think the use of the word `use` for language versioning is a Perl-ism that we should deprecate.  I would therefore propose a new statement, `language`, followed by a token indicating language *and* version:

    language raku:ver<1.0>

comes to mind.

Additionally, I think the whole language handling needs to be generalized, as to easier allow other languages to be developed in module space.  But that should probably be taken to a different issue.

**Comment by FCO** [comment]
Maybe `lang` to be consistent with `EVAL :lang`?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lizmat** [comment]
No, I mean `language`, possibly even in uppercase: `LANGUAGE`.  It should only occur once in every compilation unit, no need to huffmanize it.

**Comment by niner** [comment]
On Samstag, 21. Dezember 2019 02:13:12 CET Darren Duncan wrote:

> My preference is to actually go back to version 1.0 instead of 6.e.
> 
> Since the old versions all have letters, and if the new versions don't have
> letters, we can special case for backwards compatibility that any 6.c or
> 6.d are ordered before 1.0 or any higher numbers we eventually get to.

That would mean that the Version class would have this special casing, which 
would also apply to modules and anything else where Version objects are used. 
Or that we'd have to use a different class just for Raku versions and all 
toolchain parts would have to adopt this and keep this additional code 
forever. That's hardly good use of scarce engineering resources.

I'm also not convinced that the issue is actually worth solving. TEX does just 
fine with a versioning scheme that approaches Pi by adding digits. The very 
first version of the SUSE Linux distribution was 4.2 for obvious reasons. The 
world doesn't learn about major releases because they notice that some number 
changes, but because we put out press releases headlined "New Major Raku 
Release".

Changing the scheme does have more consequences than just a few adjustments in 
rakudo and roast. What about `use v6::inline;` to embed Raku code in Perl 5 
code?

**Comment by lizmat** [comment]
I'm not sure `Version` would need special casing: the `:ver<1.0>` is just syntactic sugar for creating a `Version` object?  The "raku" part would not be part of the `Version` object?

**Comment by eiro** [comment]
7 is even a worse idea for me because it maintains confusion with perl: perl deserve a major version for years and 7 makes perfect sense in this situation. raku is a new story so let's roll back to 3 as long as there is no technical reasons to avoid it. but someone proposed e: i love that too ... 21 major versions of raku with only one symbols could be nice and yes: e > 6

**Comment by duncand** [comment]
lizmat said:
> I think the use of the word `use` for language versioning is a Perl-ism that we should deprecate. I would therefore propose a new statement, `language`, followed by a token indicating language _and_ version:
> 
> ```
> language raku:ver<1.0>
> ```

Yes, 100% this, and the "language" written out in full.

I will note as a tangent that for a meta-declaration format I use with my Muldis Object Notation data/code interchange format I use 2 such declarations called "syntax" and "model" rather than "language" such that the first declares what literal/concept syntax or "possible representations" we have in a more abstract sense, similar to the SQL standard, while the second declares what programming language data model types etc we expect each possrep to map to, the idea being that these 2 concepts are often separable, how to parse versus what you map the parse results to.  (Technically there's also a 3rd such declaration called "script" which can be useful to explicitly disambiguate the text encoding in use etc.)  Yes, I thought about this a lot.

eiro said:
> 7 is even a worse idea for me because it maintains confusion with perl: perl deserve a major version for years and 7 makes perfect sense in this situation. raku is a new story so let's roll back to 3 as long as there is no technical reasons to avoid it. but someone proposed e: i love that too ... 21 major versions of raku with only one symbols could be nice and yes: e > 6

This has nothing to do with Perl.  We use 7 because it is numerically greater than the 6 Raku is already using.  Also Perl is at like version 30 already; if they ever change their scheme from 5.X.Y it would make sense to continue at that same X.Y and not 7 or whatever.

**Comment by eiro** [comment]
> This has nothing to do with Perl. We use 7 because it is numerically greater than the 6 Raku is already using. Also Perl is at like version 30 already; if they ever change their scheme from 5.X.Y it would make sense to continue at that same X.Y and not 7 or whatever.

so why did we use 6 ? this has something with the history of perl. if you want to brand raku as a new thing, please stop using historical schemes to decide about your versioning or your mascot.

**Comment by lizmat** [comment]
>  or your mascot

Please don't mix discussions.

**Comment by AlexDaniel** [comment]
No, we're not going to change the version backwards. If there are any changes to versioning, 7.0 is the only way it can realistically happen. It's a good thing, we can finally come up with something for minor-ish versions instead of having unversioned ever-changing errata branches (that by itself will likely require dropping unorthodox letters).

> The
world doesn't learn about major releases because they notice that some number
changes, but because we put out press releases headlined "New Major Raku
Release".

Yeah, that didn't really work when 6.c was released.

**Comment by lizmat** [comment]
Look, I understand that using letters instead of numbers was TimToady's way of working around the whole thing with the "6" in "Perl 6" being interpreted as a version indicator rather than as a language indicator.  So yes, one could go for full numerics in future versions of Raku.

On the other hand, having an "E" language indicator, does have some StarTrekky appeal.

`Raku Language Level "E" has been released.  It brings the following new features:...`

**Comment by jnthn** [comment]
> No, we're not going to change the version backwards. 

Certainly anything that happens has to have the version literals compare correctly, so the version literal of whatever new number we pick must compare greater than 6.d.

I've no problem with 7 personally, but to provide one other option for those not so keen on it: we could follow the C++ (and surely others) approach of using the year the version of the standard was released. Thus there's been C++11, C++14, C++17 - and of course, I can look at C++20 and know it's an upcoming version, because it ain't 2020 yet. So it we were to release a language version next year, it'd be Raku 20.

**Comment by jnthn** [comment]
Also, with regard to the `language` keyword proposals, I'll ponder it some, but my initial feeling is that I'm not thrilled about another keyword. I think `use` is huffmanized about right. Also - admittedly a problem that goes away in some years but worth considering - it gives a poor error for those who have a Rakudo version that doesn't know about the new keyword, where it'd just get reported as an undeclared `sub` named `language` - as opposed to the rather more helpful message that you don't have a new enough compiler.

**Comment by AlexDaniel** [comment]
> the only way it can realistically happen

> Raku 20

OK, I stand corrected, that is also an option. I'm not sure if it's better to do it that way, but it's definitely a possibility.

**Comment by taboege** [comment]
I like year-based versioning, but I wonder what the C++ Standards Committee, and Raku developers should this become the version format, will do in the 22nd century when C++(21)13 is due. C++113? (It's remarkably convenient that this is the 21st century, and we can keep year-based version strings small for some time to come.)

**Comment by niner** [comment]
No need to reach that far: C11 is already newer than C99.
I could live very well with Raku 2020 - if we decide that we need to "solve" this at all.

**Comment by duncand** [comment]
While we can use dates as a basis for version numbers, they should still be fully qualified with simple comparison rules and have longevity and account for any likely scenario.

I propose a scheme like `YYYY.X.Y.Z` that is essentially https://semver.org but that the `YYYY.X` combined is the **MAJOR** and all trailing integers are optional (then treated as zero).  How this would work is if no more than 1 version is released in a calendar year, the version is simply the 4 digits of the year, and the X/Y/Z are all implicitly zero.  If we happen to release additional MAJOR versions in the same calendar year then we increment X, first to 1, and so on.

This scheme can also be used by modules/etc if they choose, its all just a meta-spec over general version objects we already have.

**Comment by vrurg** [comment]
Use of full year would be a good option but apparently it will result in confusion with Rakudo versioning.

**Comment by duncand** [comment]
> Use of full year would be a good option but apparently it will result in confusion with Rakudo versioning.

It shouldn't cause confusion though.  Using full years for versions is a common enough generic thing.  If we're worried about confusion about this then that should be fixed in some other way, such as appropriate documentation on both projects.

What I definitely don't want to see is 2-digit year based versions, that would be horrible.

**Comment by tmtvl** [comment]
Basing versioning on the Gregorian calendar is (at least at the moment) fairly universal and, as long as we allow for an arbitrary length of year, quite future-proof.

**Comment by vrurg** [comment]
> > Use of full year would be a good option but apparently it will result in confusion with Rakudo versioning.
> 
> It shouldn't cause confusion though. Using full years for versions is a common enough generic thing. 

Try to explain to a newcomer what is the difference between _Rakudo 2020.01_ and _Raku 2020_. I personally don't see how is this better than explaining why is it _Raku 6.e_. But at least _6.e_ doesn't mix up with _2020.01_.

So far, it is only `7.x` or `YY` make real sense. The advantage of `7.x` scheme is higher flexibility in incorporating the senses of changes it would symbolize. Like `x` would stand for normal regular language releases. `x.y` for errata fixes. And `8` would mean an overhaul so big that it worth a new major.

The advantage of `YY` scheme is its simplicity and expressiveness. It is highly unlikely that two major language releases would happen consequently over the course of a single year, thus there is no need to worry about this problem. But, say, _Raku 20_ to _Raku 22_ looks distinctive enough to create real impression of changes. In this respect I'd like it more than _Raku 7.1_ to _Raku 7.2_.

Technically saying, either scheme would significantly affect Rakudo internals and MOP in particular. The reason is that whenever MOP makes a decision based on language version, the actual check is performed against revision part of the language for better performance. I.e. `Version` class is only involved once, at the beginning of compunit compilation, to provide valid parsing and handling of all possible version formats it supports. Think of all possible variants of `use v6` pragma. After that, when the exact version is already known, it is enough to `nqp::islt_s($rev, 'e')` to switch MOP execution path. This is also the reason `Metamodel::LanguageRevision` role only records revision letter to mark what compiler created a type object.

Taking into account all that, whichever scheme we choose, special care would need to be taken of how we handle the old one. Perhaps `YY` would be of the least complication as what could be done internally is `6.c` and `6.d` are mapped into 15 and 18, correspondingly (could be done at a single location without any special casing in `Version`). Then minimal changes would be required throughout the code to switch from string to numeric revision comparison.

`7.x` is worse in this respect because it would require a little more magic to prevent performance degradation when handling cross-scheme comparison. But any magic means higher level of obscurity. Non-magical handling would require involvement of `Version` with all the cost it implies.

For the bottom line, if we ever change the scheme, we must change to `YY`: `use raku:ver<20>;`

**Comment by vrurg** [comment]
> ```
> language raku:ver<1.0>
> ```

That's actually and interesting idea by itself. Perhaps it makes sense to reserve `use language` pragma for future use? For example, for speeding up some code perhaps it would make sense to provide a way of including pure NQP modules in a distribution? Then it would be done by starting a module with:

```
use language:nqp;
```

Similarly, if a module could register own compiler. Then something like the following would be possible:

```
use Language::Ruby;
use MyRubyModule;
```

_MyRubyModule.rakumod_:

```
use language:ruby;
# Ruby code goes here
```

Just a step beyond slangs.

**Comment by duncand** [comment]
Personally I hate the idea of expressing dates with 2-digit years.

However, I could support that effect if it is officially defined as COUNT of years since the Raku epoch (which is 2000 January 1 at 00:00:00 unless something changed since Larry declared that many years ago).

Since the Perl 6 project started officially in the year 2000, that means we can use a version of this form for all development versions of Raku as well, past and forward.

So for example, the Pugs project implemented preliminary Raku version 5 (note no leading zero), the old 6.c and 6.d correspond to 15 and 18, 6e could be 20, some version out in a century could be 119, and so on.

So don't call it YY, instead just X or MAJOR or whatever, and I'm okay with that.

**Comment by AlexDaniel** [comment]
@duncand yeah, cute make believe, but everyone will understand it as YY.

**Comment by vrurg** [comment]
I'm not sure if scheme naming would make it past this issue. And it undoubtedly doesn't worth bikeshedding about. 

What does matter is that:

- use of year-based language release versioning looks like a consensus
- coincidentally it is in sync with year numbers. It is a matter of personal opinion on how to interpret this fact; I don't care.
- the next language release utilizing the new scheme would most likely be Raku 22. There is a possibility for 6.e to become Raku 20, but only if the codebase is made ready for the move before Jun 2020. Even in this case `use v6.e` pragma has to be supported by the compiler.

Shall we have a vote on this?

**Comment by duncand** [comment]
Despite my general reservations for 2 digit years, I will accept the YY interpretation for versions because we're starting not before 2010 and because by the time we get to the year 2100 it will be 80 more years and likely the landscape will have completely changed by then and maintaining this scheme at that point won't need to happen.  So, pragmatism.

**Comment by niner** [comment]
I'm still not convinced that there **is** a problem to be solved here.

The original poster stated the problem as "for someone who don't know the raku history, there is no way to understand that 6.e is actually the 3rd version of the spec." But why would someone with that little knowledge about Raku even need to understand that 6.e is the 3rd version of the spec? In 20 years I've never needed to know that Perl 5.18 is the 10th major release of Perl 5 (according to one system of counting). Will anyone even try to guess how many language releases preceded C11?

Quite revealingly, not a single one of the proposed solutions will actually make it easier for someone how doesn't know the Raku history to know the how manyth release a certain version is! So I have to ask: what problem are you really trying to solve here?

As a developer, I usually ask ~ 3 questions about language versions:
- which version is higher?
- which version is the latest?
- does my compiler support a certain version?
All of which can be sufficiently answered with our current versioning scheme. Moving to a scheme based on the current year would make answering the "which version is the latest" question easier. But getting this answered has not even be stated as a problem.

So again: without knowing what we're trying to solve, throwing out solutions cannot be productive.

**Comment by duncand** [comment]
@niner  The discussion evolved from just answering the original question, and there are benefits to changing the versioning scheme that were either mentioned in later posts or were unstated but implicitly understood by many.

For me, a primary reason to change is that being able to increment the first number in a version string is valuable and normal and having the first number in a string be a fixed constant forever is typically a misfeature.

Versions in the form of dotted integer strings, such that the first integer is increased on a recurring basis, are standard practice in the industry and easy to reason about and are huffmanized.

Raku was only stuck on 6.N (and Perl on 5.N) for a long time due to historical reasons about the shared language name, and now that the name isn't shared, the versioning is free to become more normal and useful.

**Comment by vrurg** [comment]
@niner My five cents on the reasoning: consider also erratas which are currently unversioned. It's really way more of eyecandy and easier to grasp something like `Raku 20.1` (or, perhaps, `Raku 15.2` as a `6.c` alias) for errata fixes. `Raku 6.c.2`? Better not.

**Comment by niner** [comment]
Ok, I went again through all comments in this issue. The issues or benefits surrounding a version change were: 
- that the current scheme is "borderline stupid" without mentioning what's so stupid about it and how it hurts us.
- that "everyone else is still confused about the letter", with no further details or sources or how this affects us
- some undetailed "great marketing potential"
- the current scheme is "quite awkward" - again without going into any specifics

In summary, the descriptions of problems or opportunities are hand wavy at best. With the problem barely described it's gonna be hard to judge any solution.

I wouldn't care so much if the people driving the change here were the people who will do all the necessary work, but that will not be the case. As someone directly affected by any change as described above, I would hate to be forced into doing even more work than I already do for Raku just to satisfy someone's sense of aesthetics.

**Comment by vrurg** [comment]
@niner I'm not a 'driving force', rather a supporter as `YY` appeals to me but I'm also fine with the current scheme. But there is a lot of my work into the versioning too which would need to be reconsidered and redone.

As to your reason about the aesthetic nature of the discussion, well – that's what versioning is all about besides of the technical function of distinguishing different releases. People/orgs choose their schemes based on personal preferences most of the time.

The only two things which are more or less unbiased in this matter are:

- audience perception. And I anticipate `Raku 20` to be perceived somewhat more friendly than `Raku 6.e`. From media point of view it might pay back with more positive reviews in the news. That's the psychology, people pre-judge wether we like it or not and journalists are just humans.
- easier support by implementation.

   This is more essential as `YY` scheme would make some internal code simpler by taking away the need for parsing `6.X` strings in a couple of locations, starting with `use v6.X` pragma handling and continuing with CORE versions. That's not much, but still a few spared CPU cycles. Yet, even leaving those cycles aside, it would simplify the handling code and thus make it a bit more readable.

With regard to other matters, there is no rush. The change could be implemented lazily.

Otherwise that's why I propose a vote. We have enough pros and cons expressed here. So, I'm asking a 'driving force representative': can some of you put up a PR for this issue? This way could shift the discussion into more productive direction.

**Comment by coke** [comment]
I didn't see this in the comments, but my 2¢: I would avoid going directly to 7 as I think with perl5 -> 7, we're just asking for more confusion. Going to >7 is fine. (going to <6 is bad)

**Comment by jnthn** [comment]
At the point I wrote "I've no problem with 7 personally", there wasn't a Perl 7 to figure into the equation. Now there is, I think the year number is probably the best option, although ultimately this is going to come down to whether there are folks willing to do the work to make that so rather than sticking with the letters (and just dropping the `6` in marketing materials).

**Comment by vendethiel** [comment]
If year is the way to go, and considering there's a precedent for skipping a number, I suggest we skip Raku 2020..:)

**Comment by AlexDaniel** [comment]
> I think the year number is probably the best option

Is it, really? It's not semver-like and can also be a bit confusing with rakudo having the same scheme but no major version.

**Comment by AlexDaniel** [comment]
Rereading this thread, I guess the idea is to use “20”, not “2020”.

**Comment by lizmat** [comment]
If Raku wants to be a 100-year language, I cannot see how we can *not* use the full year value, aka "2020".

**Comment by AlexDaniel** [comment]
> If Raku wants to be a 100-year language, I cannot see how we can not use the full year value, aka "2020".

Maybe it's not a problem. Once it hits 99 the next version will be 100. Same as 9999 rolling to 10000, I guess. 🤷

**Comment by lizmat** [comment]
https://github.com/rakudo/rakudo/pull/4018 just shows that it should be relatively easy to allow two different version schemes in the interim.  That PR aliases `v2018` to `v6.d`, but later on it could well be the other way around, aliasing `v6.d` to `v2018`.

The PR is just there so that *if* we agree on using full year values, we can start allowing people to use them.  And the sooner we allow this, the sooner we will be able to get rid of the old (v6.letter) versioning scheme.
